File: frontend/src/types/receipt.ts
/** Тип статусу фіскалізації чеку */
export type FiscalStatus = 'none' | 'pending' | 'sent' | 'failed';

export interface Receipt {
  id: string;
  receipt_number: string;
  receipt_type: 'sale' | 'return';
  items: ReceiptItem[];
  total_amount: string;
  paid_amount?: string;
  vat_amount: string;
  payment_method: 'cash' | 'card' | 'mixed' | null;
  payment_status: PaymentStatus;
  cash_amount: string;
  card_amount: string;
  change_amount: string;
  cashier_id: string;
  cashier_name?: string;
  created_by: string;
  created_by_name?: string;
  created_at: string;
  total_profit?: number;
  original_receipt_number?: string;   // номер оригінального чеку (для повернення)
  return_reason?: string;              // причина повернення

  // ── Фіскалізація (ПРРО) ─────────────────────────────────────────────
  is_fiscal?: boolean;
  fiscal_status?: FiscalStatus | string;
  fiscal_number?: string | null;
  fiscal_serial?: string | null;
  fiscal_sent_at?: string | null;
  fiscal_error?: string | null;
  /** URL перевірки фіскального чеку (для QR-коду на друку) */
  fiscal_check_url?: string | null;

  // ── Картковий термінал (ПриватБанк) ─────────────────────────────────
  /** RRN транзакції терміналу */
  terminal_rrn?: string | null;
  /** Код авторизації (approval code) */
  terminal_approval_code?: string | null;
  /** Номер чека терміналу (invoice number) */
  terminal_invoice_number?: string | null;
  /** ID транзакції в системі терміналу */
  terminal_transaction_id?: string | null;
  /** Код відповіді терміналу (response code) */
  terminal_response_code?: string | null;
  /** Статус транзакції: 'approved' | 'declined' | ... */
  terminal_status?: string | null;
  /** Текст чека терміналу (для друку) */
  terminal_receipt?: string | null;
  /** Маскований PAN картки */
  terminal_card_pan?: string | null;
  /** Платіжна система (МПС): Visa/MasterCard/... */
  terminal_payment_system?: string | null;
  /** Мерчант (торговець) */
  terminal_merchant?: string | null;
  /** Дата/час транзакції (ISO) */
  terminal_created_at?: string | null;
}

export interface ReceiptItem {
  id: string;
  product_id: string;
  product_name: string;
  product_barcode: string | null;
  quantity: number;
  price: string;
  total: string;
  vat_rate: number;
  vat_amount: string;
  purchase_price?: number;
  profit?: number;
}

export type PaymentMethod = 'cash' | 'card' | 'mixed';
export type DebtPaymentMethod = 'cash' | 'card' | 'transfer' | 'mixed';
export type PaymentStatus = 'paid' | 'debt' | 'partially_paid';

export interface DebtPaymentInfo {
  debtor_id: string;
  amount: string;
}

export interface ReceiptCreate {
  receipt_number?: string;
  receipt_type: 'sale' | 'return';
  cashier_id?: string;
  cashier_name?: string;
  total_amount: string;
  paid_amount?: string;
  debtor_id?: string;
  items: ReceiptItemCreate[];
  payment_method?: PaymentMethod;
  cash_amount?: number;
  card_amount?: number;
  is_debt?: boolean;
  debt_payment_method?: string;
  debt_payment?: DebtPaymentInfo;
  original_receipt_id?: string;  // ID оригінального чеку (для повернення)
  return_reason?: string;         // причина повернення

  // ── Картковий термінал (ПриватБанк) ─────────────────────────────────
  terminal_rrn?: string | null;
  terminal_approval_code?: string | null;
  terminal_invoice_number?: string | null;
  terminal_transaction_id?: string | null;
  terminal_response_code?: string | null;
  terminal_status?: string | null;
  terminal_receipt?: string | null;
  terminal_card_pan?: string | null;
  terminal_payment_system?: string | null;
  terminal_merchant?: string | null;
  terminal_created_at?: string | null;
}

export interface ReceiptItemCreate {
  product_id: string;
  quantity: number;
  price: number;
}

// ── Для пошуку чеків ──────────────────────────
export interface ReceiptSearchResult {
  id: string;
  receipt_number: string;
  receipt_type: 'sale' | 'return';
  total_amount: number;
  created_at: string;
  cashier_name: string;
  items_count: number;
}

// ── Для повернення без чеку ────────────────────
export interface RecentSaleInfo {
  receipt_id: string;
  receipt_number: string;
  created_at: string;
  quantity: number;
  price: number;
}

export interface ProductRecentSalesResponse {
  product: {
    id: string;
    title: string;
    barcode: string | null;
    price: number;
    unit: string;
  };
  total_sold: number;
  total_returned: number;
  returnable: number;
  recent_sales: RecentSaleInfo[];
}

export interface ProductRecentSalesListResponse {
  items: ProductRecentSalesResponse[];
  total: number;
}

/** Зручні функції для статусу фіскалізації */
export const FISCAL_STATUS_LABELS: Record<string, string> = {
  none: 'Не фіскалізовано',
  pending: 'Очікує',
  sent: 'Фіскалізовано',
  failed: 'Помилка',
};

export function getFiscalStatusLabel(status?: string | null): string {
  if (!status) return FISCAL_STATUS_LABELS.none;
  return FISCAL_STATUS_LABELS[status] || status;
}

File: backend/app/schemas/receipt.py
"""
Pydantic схеми для моделей Receipt та ReceiptItem (Чек продажу / повернення).
"""

from decimal import Decimal
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, ConfigDict

from app.models.receipt import ReceiptType, ReceiptPaymentMethod


class DebtPaymentInfo(BaseModel):
    """Інформація про оплату боргу через касу."""
    debtor_id: UUID = Field(..., description="ID боржника")
    amount: Decimal = Field(..., gt=0, max_digits=12, decimal_places=2, description="Сума оплати боргу")


class ReceiptItemCreate(BaseModel):
    """Схема створення позиції чеку."""
    product_id: UUID = Field(..., description="ID товару")
    quantity: Decimal = Field(..., max_digits=10, decimal_places=3, description="Кількість")
    price: Decimal = Field(..., max_digits=10, decimal_places=2, description="Ціна за одиницю (грн)")
    total: Optional[Decimal] = Field(None, max_digits=12, decimal_places=2, description="Загальна сума (грн)")


class ReceiptItemResponse(BaseModel):
    """Схема відповіді з даними позиції чеку."""
    id: UUID
    receipt_id: UUID
    product_id: UUID
    product_name: str = Field("", description="Назва товару")
    product_barcode: Optional[str] = Field(None, description="Штрих-код товару")
    quantity: Decimal
    price: Decimal
    total: Decimal
    purchase_price: Optional[Decimal] = Field(None, description="Собівартість товару на момент продажу")
    profit: Optional[Decimal] = Field(None, description="Прибуток = total - (purchase_price * quantity)")
    vat_amount: Optional[Decimal] = Field(None, description="Сума ПДВ для цієї позиції")
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ReceiptCreate(BaseModel):
    """Схема створення нового чеку (продаж або повернення)."""
    receipt_number: Optional[str] = Field(None, max_length=50, description="Номер чеку")
    receipt_type: ReceiptType = Field(ReceiptType.SALE, description="Тип чеку: sale або return")
    cashier_id: Optional[UUID] = Field(None, description="ID касира")
    total_amount: Decimal = Field(..., max_digits=12, decimal_places=2, description="Загальна сума чеку (грн)")
    paid_amount: Optional[Decimal] = Field(None, max_digits=12, decimal_places=2, description="Фактично сплачена сума (грн)")
    debtor_id: Optional[UUID] = Field(None, description="ID боржника (якщо покупка в борг)")
    is_return: bool = Field(False, description="Чи є поверненням")
    notes: Optional[str] = Field(None, description="Нотатки до чеку")
    original_receipt_id: Optional[UUID] = Field(None, description="ID оригінального чеку (для повернень)")
    return_reason: Optional[str] = Field(None, max_length=500, description="Причина повернення")
    items: list[ReceiptItemCreate] = Field(default_factory=list, description="Позиції чеку")
    debt_payment: Optional[DebtPaymentInfo] = Field(None, description="Оплата боргу через касу")
    payment_method: Optional[ReceiptPaymentMethod] = Field(None, description="Спосіб оплати: cash, card, mixed")


class ReceiptResponse(BaseModel):
    """Схема відповіді з даними чеку."""
    id: UUID
    receipt_number: str
    receipt_type: ReceiptType
    cashier_id: UUID
    total_amount: Decimal
    paid_amount: Optional[Decimal] = Field(None, description="Фактично сплачена сума")
    change_amount: Optional[Decimal] = Field(None, description="Сума здачі (грн)")
    debtor_id: Optional[UUID] = Field(None, description="ID боржника")
    is_return: bool
    notes: Optional[str] = None
    created_at: datetime
    items: list[ReceiptItemResponse] = []
    total_profit: Optional[Decimal] = Field(None, description="Загальний чистий прибуток по чеку")
    vat_amount: Optional[Decimal] = Field(None, description="Загальна сума ПДВ по чеку")
    cashier_name: str = Field("", description="Ім'я касира")
    payment_method: Optional[str] = Field(None, description="Спосіб оплати: cash, card, mixed")

    model_config = ConfigDict(from_attributes=True)


class ReceiptHistoryParams(BaseModel):
    """Параметри для перегляду історії чеків."""
    cashier_id: Optional[UUID] = Field(None, description="Фільтр за касиром")
    receipt_type: Optional[ReceiptType] = Field(None, description="Фільтр за типом")
    date_from: Optional[datetime] = Field(None, description="Початкова дата")
    date_to: Optional[datetime] = Field(None, description="Кінцева дата")
    page: int = Field(1, ge=1, description="Номер сторінки")
    size: int = Field(20, ge=1, le=100, description="Розмір сторінки")


# ─── Схеми для повернення товарів ───────────────────────────────

class ReceiptSearchResult(BaseModel):
    """Результат пошуку чеків для повернення."""
    id: UUID
    receipt_number: str
    receipt_type: ReceiptType
    total_amount: Decimal
    created_at: datetime
    cashier_name: str = ""
    items_count: int = 0


class ProductBriefInfo(BaseModel):
    """Коротка інформація про товар для повернення."""
    id: UUID
    title: str
    barcode: Optional[str] = None
    price: Decimal
    unit: str


class RecentSaleInfo(BaseModel):
    """Інформація про один продаж товару (для повернення без чеку)."""
    receipt_id: UUID
    receipt_number: str
    created_at: datetime
    quantity: Decimal
    price: Decimal


class ProductRecentSalesResponse(BaseModel):
    """Відповідь з інформацією про останні продажі товару."""
    product: ProductBriefInfo
    total_sold: Decimal
    total_returned: Decimal
    returnable: Decimal
    recent_sales: list[RecentSaleInfo]


class ProductRecentSalesListResponse(BaseModel):
    """Відповідь зі списком товарів, знайдених за запитом."""
    items: list[ProductRecentSalesResponse]
    total: int
